ssl/t1_lib.c

```diff
@@ -623,43 +624,35 @@ static int add_provider_sigalgs(const OSSL_PARAM params[], void *data)
          * behaviour)?
          */
         if (EVP_KEYMGMT_get0_provider(keymgmt) == provider) {
-            int keytype_nid;
             /* We have a match - so we could use this signature */
             /* Check proper object registration first, though */
             /* don't care about return value as this may have been
              * done within providers or previous calls to
              * add_provider_sigalgs
              */
             OBJ_create(sinf->sigalg_oid, sinf->sigalg_name, NULL);
-	    if (sinf->sig_name)
+            if (sinf->sig_name != NULL)
                 OBJ_create(sinf->sig_oid, sinf->sig_name, NULL);
-	    if (sinf->keytype) {
-                keytype_nid = OBJ_create(sinf->keytype_oid, sinf->keytype, NULL);
-            }
-            else {
-                if (sinf->sig_name)
-                    keytype_nid = OBJ_txt2nid(sinf->sig_name);
-                else
-                    keytype_nid = OBJ_txt2nid(sinf->sigalg_name);
-            }
-	    if (sinf->hash_name) {
+            if (sinf->keytype != NULL)
+                OBJ_create(sinf->keytype_oid, sinf->keytype, NULL);
+	    if (sinf->hash_name != NULL) {
                 OBJ_create(sinf->hash_oid, sinf->hash_name, NULL);
-                if (sinf->sig_name)
-                    OBJ_add_sigid(OBJ_txt2nid(sinf->sig_name),
+                if (sinf->sig_name != NULL)
+                    OBJ_add_sigid(OBJ_txt2nid(sinf->sigalg_name),
                                   OBJ_txt2nid(sinf->hash_name),
-                                  keytype_nid);
-            }
-	    else {
+                                  OBJ_txt2nid(sinf->sig_name));
+            } else {
                 OBJ_add_sigid(OBJ_txt2nid(sinf->sigalg_name),
                               NID_undef,
-                              keytype_nid);
+                              (sinf->sig_name == NULL)?NID_undef:
+                                             OBJ_txt2nid(sinf->sig_name));
```

Comment:
For the case where `sinf->sig_name == NULL`, this call makes no sense at all (and will return an error).
I would rather not have to add a sigid sigalg that's exactly the same as the sig, but I assume there's other code that relies on the presence of sigids even though not necessary.

So my suggestion is rather this:

```suggestion
                OBJ_add_sigid(OBJ_txt2nid(sinf->sigalg_name),
                              NID_undef,
                              (sinf->sig_name != NULL)
                              ? OBJ_txt2nid(sinf->sig_name)
                              : OBJ_txt2nid(sinf->sigalg_name));
```